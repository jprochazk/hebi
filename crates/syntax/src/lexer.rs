#![allow(non_camel_case_types)]

use std::fmt;

use logos::Logos;
use span::Span;

#[derive(Clone, Debug)]
pub struct Token {
  pub ws: usize,
  pub kind: TokenKind,
  pub span: Span,
}

pub struct Lexer<'src> {
  src: &'src str,
  tokens: Vec<Token>,
  eof: Span,
}

#[derive(Debug)]
pub struct Error {
  pub span: Span,
  pub lexeme: String,
}

impl<'src> Lexer<'src> {
  pub fn lex(src: &'src str) -> Result<Self, Vec<Error>> {
    let eof = (src.len()..src.len()).into();

    let mut ws = 0;
    let mut errors = vec![];
    let mut tokens = vec![];
    let mut lexer = logos::Lexer::<'src, TokenKind>::new(src);
    while let Some(kind) = lexer.next() {
      let lexeme = lexer.slice();
      let span = lexer.span().into();

      match kind {
        // Handle whitespace
        TokenKind::_Indentation => {
          ws = lexeme.trim_start_matches(|c| c == '\n' || c == '\r').len();
          continue;
        }
        // Filter any other whitespace and comments
        TokenKind::_Whitespace | TokenKind::_Comment => continue,
        // Handle errors
        TokenKind::_Error => {
          errors.push(Error {
            lexeme: lexeme.into(),
            span,
          });
          continue;
        }
        // Any other token is stored with its preceding whitespace
        _ => {
          tokens.push(Token { ws, kind, span });
          ws = 0;
        }
      }
    }

    if !errors.is_empty() {
      Err(errors)
    } else {
      Ok(Lexer { src, tokens, eof })
    }
  }

  fn lexeme(&self, span: Span) -> &'src str {
    &self.src[span.start..span.end]
  }
}

impl<'src> peg::Parse for Lexer<'src> {
  type PositionRepr = Span;

  fn start(&self) -> usize {
    0
  }

  fn is_eof(&self, pos: usize) -> bool {
    pos >= self.tokens.len()
  }

  fn position_repr(&self, pos: usize) -> Self::PositionRepr {
    self
      .tokens
      .get(pos)
      .map(|t| t.span)
      .unwrap_or_else(|| self.eof)
  }
}

impl<'src> peg::ParseElem<'src> for Lexer<'src> {
  type Element = &'src Token;

  fn parse_elem(&'src self, pos: usize) -> peg::RuleResult<Self::Element> {
    match self.tokens.get(pos) {
      Some(token) => peg::RuleResult::Matched(pos + 1, token),
      None => peg::RuleResult::Failed,
    }
  }
}

impl<'src> peg::ParseLiteral for Lexer<'src> {
  fn parse_string_literal(&self, pos: usize, literal: &str) -> peg::RuleResult<()> {
    let Some(token) = self.tokens.get(pos) else {
      return peg::RuleResult::Failed;
    };

    if self.lexeme(token.span) == literal {
      peg::RuleResult::Matched(pos + 1, ())
    } else {
      peg::RuleResult::Failed
    }
  }
}

impl<'src> peg::ParseSlice<'src> for Lexer<'src> {
  type Slice = &'src [Token];

  fn parse_slice(&'src self, p1: usize, p2: usize) -> Self::Slice {
    &self.tokens[p1..p2]
  }
}

#[derive(Clone, Debug, Logos)]
pub enum TokenKind {
  // Keywords
  #[token("use")]
  Kw_Use,
  #[token("pub")]
  Kw_Pub,
  #[token("fn")]
  Kw_Fn,
  #[token("yield")]
  Kw_Yield,
  #[token("class")]
  Kw_Class,
  #[token("for")]
  Kw_For,
  #[token("in")]
  Kw_In,
  #[token("while")]
  Kw_While,
  #[token("loop")]
  Kw_Loop,
  #[token("return")]
  Kw_Return,
  #[token("break")]
  Kw_Break,
  #[token("continue")]
  Kw_Continue,
  #[token("if")]
  Kw_If,
  #[token("elif")]
  Kw_Elif,
  #[token("else")]
  Kw_Else,

  // Brackets
  #[token("{")]
  Brk_CurlyL,
  #[token("}")]
  Brk_CurlyR,
  #[token("(")]
  Brk_ParenL,
  #[token(")")]
  Brk_ParenR,
  #[token("[")]
  Brk_SquareL,
  #[token("]")]
  Brk_SquareR,

  // Misc characters
  #[token(".")]
  Op_Dot,
  #[token(",")]
  Comma,
  #[token(";")]
  Semicolon,
  #[token(":")]
  Colon,
  #[token("?")]
  Question,

  // Equals operators
  #[token("=")]
  Op_Equal,
  #[token("==")]
  Op_EqualEqual,
  #[token("+=")]
  Op_PlusEqual,
  #[token("-=")]
  Op_MinusEqual,
  #[token("/=")]
  Op_SlashEqual,
  #[token("*=")]
  Op_StarEqual,
  #[token("%=")]
  Op_PercentEqual,
  #[token("**=")]
  Op_StarStarEqual,
  #[token("??=")]
  Op_QuestionQuestionEqual,
  #[token("!=")]
  Op_BangEqual,

  // Operators
  #[token("+")]
  Op_Plus,
  #[token("-")]
  Op_Minus,
  #[token("/")]
  Op_Slash,
  #[token("*")]
  Op_Star,
  #[token("%")]
  Op_Percent,
  #[token("**")]
  Op_StarStar,
  #[token("??")]
  Op_QuestionQuestion,
  #[token("!")]
  Op_Bang,
  #[token(">")]
  Op_More,
  #[token(">=")]
  Op_MoreEqual,
  #[token("<")]
  Op_Less,
  #[token("<=")]
  Op_LessEqual,
  #[token("||")]
  Op_PipePipe,
  #[token("&&")]
  Op_AndAnd,
  #[token("..")]
  Op_Range,
  #[token("..=")]
  Op_RangeInc,

  // Literals
  /// `null`
  #[token("null")]
  Lit_Null,
  /// `0`, `1.0`, `5e10`, etc.
  #[regex(r"[0-9]+(\.[0-9]+)?([Ee][+-]?[0-9]+)?")]
  Lit_Number,
  /// `true` or `false`
  #[token("true")]
  #[token("false")]
  Lit_Bool,
  #[regex(r#""([^"\\]|\\.)*""#)]
  Lit_String,
  /// `a`, `b_c`, `__x0`, etc.
  #[regex("[a-zA-Z_][a-zA-Z0-9_]*")]
  Lit_Ident,

  #[doc(hidden)]
  #[regex(r"\n\r?[ ]+")]
  _Indentation,
  #[doc(hidden)]
  #[regex(r"[ \n\r]+", logos::skip)]
  _Whitespace,
  #[doc(hidden)]
  #[regex(r"#[^\n]*", logos::skip)]
  _Comment,

  /// Errors are filtered out before parsing
  #[doc(hidden)]
  #[error]
  _Error,
}

impl<'src> fmt::Debug for Lexer<'src> {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    struct DebugToken<'src>(&'src str, Token);
    impl<'src> fmt::Debug for DebugToken<'src> {
      fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if let TokenKind::Lit_Ident = self.1.kind {
          write!(
            f,
            "(>{} {:?} `{}` @{})",
            self.1.ws, self.1.kind, self.0, self.1.span
          )
        } else {
          write!(f, "(>{} {:?} @{})", self.1.ws, self.1.kind, self.1.span)
        }
      }
    }

    self
      .tokens
      .clone()
      .into_iter()
      .map(|token| DebugToken(self.lexeme(token.span), token))
      .collect::<Vec<_>>()
      .fmt(f)
  }
}

#[cfg(test)]
mod tests;
