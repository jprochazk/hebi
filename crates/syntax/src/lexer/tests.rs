use super::*;

#[test]
fn lex_design_file() {
  let input = r#"
# values
v = none # none
v = 0.1 # number
v = true # bool
v = "\tas\\df\x2800\n" # string
v = [none, 0.1, true, "\tas\\df\x2800\n"] # array
# object
v = {a=none, b=0.1, c=true, d="\tas\\df\x2800\n"}
v = {["a"]=none, ["b"]=0.1, ["c"]=true, ["d"]="\tas\\df\x2800\n"}
v = {[0]=none, [1]=0.1, [2]=true, [3]="\tas\\df\x2800\n"}

# operators
v = 2 + 2
v = 2 - 2
v = 2 / 2
v = 2 * 2
v = 2 % 2
v = 2 ** 2
v = 2 == 2
v = 2 != 2
v = 2 > 2
v = 2 >= 2
v = 2 < 2
v = 2 <= 2
v = -2
v = !true
v = true && true
v = false || true
v = a ?? b

# assignment
v = 1 # first ocurrence in a scope creates a new variable
v += 1
v -= 1
v /= 1
v *= 1
v %= 1
v **= 1
v ??= 1

# postfix
v.a
v["a"]
v(a)

# functions
# last expr return
fn add(a, b):
  a + b

v = add(0, 1)

fn fib(n):
  if n < 2: n
  else: n * fib(n - 1)
fn print_fib(n):
  print(fib(n))

# loops
# range is an object
for i in 0..10:
  print(i)

# `yield` inside `fn` makes it an iterator
# when called, iterators return an object with a `next` method
# an iterator is done when its `next` method returns none
fn counter(start, step, end):
  n = start
  loop:
    yield n
    n += step
    if end && n > end: return

for n in counter(0, 10, 100):
  print(n)

v = 0
while v < 10:
  print(v)
  v += 1

v = 0
loop:
  if v >= 10: break
  print(v)
  v += 1

if v >= 10: print("larger than 10", v)
else: print("not larger than 10", v)

if v < 10:
  print("less than 10")
else if v < 20:
  print("less than 20")
else:
  print("very large")

# `if` is an expression
fn clamp(v, min, max):
  if v < min: min
  elif v > max: max
  else: v

class Test:
  init(self, n):
    self.n = n

  get_n(self):
    self.n

  test0():
    print("static", Test)

  test1(self):
    print("instance", self)

v = Test()
print(v.get_n() == Test.get_n(v)) # true

# errors
# no exceptions, panic = abort
panic("asdf")

# modules
# json_test.t
import json
# other ways to import:
# import json.parse
# import json.{parse}
# import {json}
# import {json.parse}
# import {json.{parse}}

v = json.parse("{\"a\":0, \"b\":1}")
print(v) # { a: 0, b: 1 }
"#;

  let lexer = Lexer::new(input);
  let tokens = Tokens(lexer)
    .map(|(s, t)| DebugToken(t, s))
    .collect::<Vec<_>>();

  insta::assert_debug_snapshot!(tokens)
}
