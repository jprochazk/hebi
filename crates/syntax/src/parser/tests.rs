use indoc::indoc;

use super::*;
use crate::lexer::Lexer;

// TODO: emit input expression in snapshot
// do this for all snapshots tests that don't do it already

macro_rules! check_module {
  ($input:literal) => {
    let input = indoc!($input);
    match parse(input) {
      Ok(module) => insta::assert_debug_snapshot!(module),
      Err(e) => {
        for err in e {
          eprintln!("{}", err.report(input));
        }
        panic!("Failed to parse source, see errors above.")
      }
    };
  };
}

macro_rules! check_expr {
  ($input:literal) => {
    let input = $input;
    match Parser::new(Lexer::new(input)).expr() {
      Ok(module) => insta::assert_debug_snapshot!(module),
      Err(err) => {
        eprintln!("{}", err.report(input));
        panic!("Failed to parse source, see errors above.")
      }
    };
  };
}

macro_rules! check_error {
  ($input:literal) => {
    let input = indoc!($input);
    match parse(input) {
      Ok(_) => panic!("module parsed successfully"),
      Err(e) => {
        let mut errors = String::new();
        for err in e {
          errors += &err.report(input);
          errors += "\n";
        }
        insta::assert_snapshot!(errors);
      }
    };
  };
}

#[test]
fn import_stmt() {
  check_module! {
    r#"
      import a
      import a.b
      import a.b.c
      import a.{b, c}
      import a.{b.{c}, d.{e}}
      import {a.{b}, c.{d}}
      import {a, b, c,}
    "#
  };

  check_module! {
    r#"
      import a as x
      import a.b as x
      import a.b.c as x
      import a.{b as x, c as y}
      import a.{b.{c as x}, d.{e as y}}
      import {a.{b as x}, c.{d as y}}
      import {a as x, b as y, c as z,}
    "#
  };

  check_error! {
    r#"
      import a
        import b
    "#
  };
}

#[test]
fn binary_expr() {
  check_expr!(r#"a + b"#);
  check_expr!(r#"a - b"#);
  check_expr!(r#"a / b"#);
  check_expr!(r#"a ** b"#);
  check_expr!(r#"a * b"#);
  check_expr!(r#"a % b"#);
  check_expr!(r#"a == b"#);
  check_expr!(r#"a != b"#);
  check_expr!(r#"a > b"#);
  check_expr!(r#"a >= b"#);
  check_expr!(r#"a < b"#);
  check_expr!(r#"a <= b"#);
  check_expr!(r#"a && b"#);
  check_expr!(r#"a || b"#);
  check_expr!(r#"a ?? b"#);

  check_module! {
    r#"
      a + b
      c + d
    "#
  };

  check_error! {
    r#"
      a +
        b
    "#
  }
}

#[test]
fn unary_expr() {
  check_expr!(r#"+a"#);
  check_expr!(r#"-a"#);
  check_expr!(r#"!a"#);
  check_expr!(r#"?a.b[c].d()"#);
}

#[test]
fn postfix_expr() {
  check_expr!(r#"a.b[c].d"#);
  check_module! {
    r#"
      a.b[c].d
      a.b[c].d
    "#
  };

  check_error! {
    r#"
      a
      .b[c].d
    "#
  }
  check_error! {
    r#"
      a.b[c]
      .d
    "#
  }
  check_error! {
    r#"a."#
  }
}

#[test]
fn call_expr() {
  check_expr!(r#"a(b, c, d=e, f=g)"#);
  check_module! {
    r#"
      a(b, c, d=e, f=g)
      a(
        b,
      c, d
          =e,
        f=
        g,
          )
    "#
  };

  check_error! {
    r#"
      a(b=c, d)
    "#
  }
}

#[test]
fn simple_literal_expr() {
  check_module! {
    r#"
      none
      true
      false
      1
      0.1
      1.5e3
      3.14e-3
      "\tas\\df\u{2800}\x28\n"
    "#
  }
}

#[test]
fn array_literal_expr() {
  check_module! {
    r#"
      [0, 1, 2]
      [0,
       1,
       2,]
      [
        0,
        1,
        2,
      ]
    "#
  }
}

#[test]
fn dict_literal_expr() {
  check_module! {
    r#"
      {a:b, c:d, e:f}
      {a:b,
        c:d,
        e:f,}
      {
        a:b,
        c:d,
        e:f,
      }
    "#
  }

  check_module! {
    r#"
      {[a]:b, [c]:d, [e]:f}
      {[a]:b,
       [c]:d,
       [e]:f,}
      {
        [a]:b,
        [c]:d,
        [e]:f,
      }
    "#
  }
}

#[test]
fn grouping_expr() {
  check_module! {
    r#"
      # asdf
      (a + b)
      (
      a
       +
          b
      )
      (a
        (b))
      ((((a))))
    "#
  }
}

#[test]
fn assign_expr() {
  check_module! {
    r#"
      # asdf
      a = b
      a := b
      a += b
      a -= b
      a /= b
      a *= b
      a %= b
      a **= b
      a ??= b
    "#
  }

  check_module! {
    r#"
      # asdf
      a.b = b
      a.b += b
      a.b -= b
      a.b /= b
      a.b *= b
      a.b %= b
      a.b **= b
      a.b ??= b
    "#
  }

  check_error! {
    r#"
      a
        = b
    "#
  }
  check_error! {
    r#"
      a =
        b
    "#
  }

  check_error! {
    r#"a.b := c"#
  }

  check_error! {
    r#"a() = b"#
  }
}

#[test]
fn if_stmt() {
  check_module! {
    r#"
      if a: pass
      elif b: pass
      elif c: pass
      else: pass
    "#
  }

  check_module! {
    r#"
      if a:
        pass
      elif b:
        pass
      elif c:
        pass
      else:
        pass
    "#
  }

  check_module! {
    r#"
      if a:
        a
        b
      elif b:
        a
        b
      elif c:
        a
        b
      else:
        a
        b
    "#
  }

  check_module! {
    r#"
      if a:
        if b:
          pass
    "#
  }

  check_error! {
    r#"
      if a:
        a
          b
      else: pass
    "#
  }

  check_error! {
    r#"
      if a
        : pass
      else: pass
    "#
  }

  check_error! {
    r#"
      if a: pass
      elif b
        : pass
      else: pass
    "#
  }

  check_error! {
    r#"
      if a: pass
      elif b: pass
      else
        : pass
    "#
  }

  check_error! {
    r#"
      if a: pass
      elif b: pass
        else: pass
    "#
  }

  check_error! {
    r#"
      if a: pass
        elif b: pass
      else: pass
    "#
  }

  check_module! {
    r#"
      if (
        some_very_long_condition &&
        with_many_sub_expressions() ||
        (which_are_complex.too())
      ):
        pass
    "#
  }
}

#[test]
fn loop_stmts() {
  check_module! {
    r#"
      loop: pass
      loop:
        pass
    "#
  }

  check_error! {
    r#"
      loop:
      pass
    "#
  }

  check_module! {
    r#"
      while true: pass
      while true:
        pass
    "#
  }

  check_error! {
    r#"
      while true:
      pass
    "#
  }

  check_module! {
    r#"
      for i in iter(): pass
      for i in iter():
        pass
      for i in 0..10: pass
      for i in 0..10:
        pass
      for i in a()..b(): pass
      for i in a()..b():
        pass
      for i in 0..=10: pass
      for i in 0..=10:
        pass
      for i in a()..=b(): pass
      for i in a()..=b():
        pass
    "#
  }

  check_error! {
    r#"
      for i in iter():
      pass
    "#
  }

  check_module! {
    r#"
      loop:
        loop:
          a
          a
    "#
  }
}

#[test]
fn func_stmt() {
  check_module! {
    r#"
      fn f(a, b, c,): pass
      fn f(a, b, c=d): pass
      fn f(*argv,): pass
      fn f(**kwargs,): pass
      fn f(*argv, **kwargs,): pass
      fn f(a, b=c, *argv, d=f, g, **kwargs,): pass
    "#
  }

  check_error!(r#"fn f(a, b=c, d,): pass"#);
  check_error!(r#"fn f(*,): pass"#);
  check_error!(r#"fn f(**,): pass"#);
  check_error!(r#"fn f(**kwargs, a,): pass"#);
  check_error!(r#"fn f(a, b=,): pass"#);
  check_error!(r#"fn f(a, a): pass"#);
  check_error!(r#"fn f(a, *a): pass"#);
  check_error!(r#"fn f(a, *, a): pass"#);
  check_error!(r#"fn f(a, **a): pass"#);
  check_error! {
    r#"
      fn f():
      pass
    "#
  }
  check_error! {
    r#"
      fn():
        pass
    "#
  }
}

#[test]
fn ctrl_stmt() {
  check_module! {
    r#"
      # simple
      loop:
        break
        continue
      
      fn f():
        v = yield
        v = yield v
        yield
        yield v
        return
        return v

      # nested
      loop:
        loop:
          break
          continue
        break
        continue

      fn g():
        fn h():
          v = yield
          v = yield v
          yield
          yield v
          return
          return v
        yield
        yield v
        return
        return v
      
      loop:
        fn i():
          v = yield
          v = yield v
          yield
          yield v
          return
          return v
        break
        continue

      fn j():
        loop:
          break
          continue
        v = yield
        v = yield v
        yield
        yield v
        return
        return v
      
      loop:
        fn k():
          loop:
            break
            continue
          v = yield
          v = yield v
          yield
          yield v
          return
          return v
        break
        continue

      fn l():
        loop:
          fn m():
            v = yield
            v = yield v
            yield
            yield v
            return
            return v
          break
          continue
        v = yield
        v = yield v
        yield
        yield v
        return
        return v
    "#
  }

  check_error! {
    r#"
      return v
    "#
  }

  check_error! {
    r#"
      yield v
    "#
  }

  check_error! {
    r#"
      continue
    "#
  }

  check_error! {
    r#"
      break
    "#
  }
}

#[test]
fn print_stmt() {
  check_module! {
    r#"
      print "a", 0, true
      print "a", 0, true
    "#
  }

  check_error! {
    r#"
      print
        "a"
    "#
  }
}

#[test]
fn class_stmt() {
  check_module! {
    r#"
      class T: pass
      class T:
        pass
      class T:
        fn f(v): pass
      class T:
        a = b
        fn f(v): pass
      class T(U): pass
      class T(U):
        pass
      class T(U):
        a = b
      class T(U):
        a = b
        fn f(v): pass
      class T(U):
        a = b
        fn f(v):
          pass
      class T:
        a
      class T(U):
        a
    "#
  }

  check_error! {
    r#"
      class
        T: pass
    "#
  }

  check_error! {
    r#"
      class T
        : pass
    "#
  }

  check_error! {
    r#"
      class T: a = b
    "#
  }

  check_error! {
    r#"
      class T: fn f(v): pass
    "#
  }

  check_error! {
    r#"
      class T
        (U): pass
    "#
  }

  check_error! {
    r#"
      class T(U)
        : pass
    "#
  }

  check_error! {
    r#"
      class T(U):
        fn f(v): pass
        a = b
    "#
  }
}

#[test]
fn class_self_and_super() {
  check_module! {
    r#"
      class T:
        fn f(self):
          print self
      
      class T(U):
        fn f(self):
          print self, super

      class T(U):
        v = super.f()
    "#
  }

  check_error! {
    r#"
      self
    "#
  }

  check_error! {
    r#"
      class T:
        v = self.f()
    "#
  }

  check_error! {
    r#"
      fn f():
        print self
    "#
  }

  check_error! {
    r#"
      class T:
        fn f():
          print self
    "#
  }

  check_error! {
    r#"
      super
    "#
  }

  check_error! {
    r#"
      fn f():
        print super
    "#
  }

  check_error! {
    r#"
      class T:
        fn f():
          print self
    "#
  }
}

#[test]
fn whole_module() {
  check_module! {
    r#"
      # variable declaration
      v := 0

      # values
      v = none # none
      v = 0.1 # number
      v = true # bool
      v = "\tas\\df\x2800\n" # string
      v = [none, 0.1, true, "\tas\\df\x2800\n"] # list
      v = {a: none, b: 0.1, c: true, d: "\tas\\df\x2800\n"} # dict
      v = {["a"]: none, ["b"]: 0.1, ["c"]: true, ["d"]: "\tas\\df\x2800\n"}
      v = {[0]: none, [1]: 0.1, [2]: true, [3]: "\tas\\df\x2800\n"}

      # operators
      v = 2 + 2
      v = 2 - 2
      v = 2 / 2
      v = 2 * 2
      v = 2 % 2
      v = 2 ** 2
      v = 2 == 2
      v = 2 != 2
      v = 2 > 2
      v = 2 >= 2
      v = 2 < 2
      v = 2 <= 2
      v = -2
      v = !true
      v = true && true
      v = false || true
      v = a ?? b

      # assignment
      v = 1
      v += 1
      v -= 1
      v /= 1
      v *= 1
      v %= 1
      v **= 1
      v ??= 1

      # postfix
      v.a
      v["a"]
      v(a)

      # functions
      fn add(a, b):
        return a + b

      v = add(0, 1)

      fn fact(n):
        if n < 2:
          return n
        else:
          return n * fact(n - 1)

      fn print_fact(n):
        print(fact(n))

      # loops
      # range is an object
      for i in 0..10:
        print(i)

      # `yield` inside `fn` makes it an iterator
      # when called, iterators return an object with a `next` method
      # an iterator is done when its `next` method returns none
      fn counter(start, step, end):
        n := start
        loop:
          yield n
          n += step
          if end && n > end:
            return

      for n in counter(0, 10, 100):
        print(n)

      v = 0
      while v < 10:
        print(v)
        v += 1

      v = 0
      loop:
        if v >= 10:
          break
        print(v)
        v += 1

      if v < 10:
        print("less than 10")
      elif v < 20:
        print("less than 20")
      else:
        print("very large")

      class Test:
        fn init(self, n):
          self.n = n

        fn get_n(self):
          return self.n

        fn test1(self):
          print("instance", self)

        fn test0():
          print("static", Test)

      v = Test()
      print(v.get_n() == Test.get_n(v)) # true

      v = Test(n=10)

      Test.test0()
      v.test1()

      # errors
      # no exceptions, panic = abort
      panic("asdf")

      # modules
      # json_test.t
      import json
      # other ways to import:
      # import json.parse
      # import json.{parse}
      # import {json}
      # import {json.parse}
      # import {json.{parse}}

      v = json.parse("{\"a\":0, \"b\":1}")
      print(v) # { a: 0, b: 1 }

      # data class, implicit initializer
      class A:
        a = 100
        # fn init(self, a = 100):
        #   self.a = a

      print(A().a)     # 100
      print(A(a=10).a) # 10

      class B:
        a = 100
        fn init(self): # override the implicit initializer
          pass

      print(B().a)   # 100
      # `a` is ignored
      print(B(a=10)) # 100

      class C:
        # fields do not have to be declared
        # and may be added in the initializer
        # after `init` is called, the class is frozen
        # no fields/methods may be added or removed
        fn init(self):
          self.a = 10

      print(C().a) # 10
      C().b = 10 # error: cannot add new field `b` to class `C`

      class A:
        fn inherited(self):
          print("test 0")

      class B(A): pass

      A().inherited() # test 0
      B().inherited() # test 0

      class C(B):
        fn inherited(self): # override
          print("test 1")

      C().inherited() # test 1

      class D(C):
        fn inherited(self): # override with call to super
          super.inherited()
          print("test 2")

      D().inherited() # test 1
                      # test 2

      class X:
       fn init(self):
          self.v = 10

      class Y(X):
        fn init(self): # error: `super.init` must be called before accessing `self` or returning in derived constructor
          self.v = 10

      class Z(X):
        fn init(self, v):
          super.init(self)
          self.v += v

      print(Z(v=15).v) # 25
    "#
  }
}

/* #[test]
fn _temp() {
  check_error! {
    r#"
      class T: fn f(v): pass
    "#
  }
} */
