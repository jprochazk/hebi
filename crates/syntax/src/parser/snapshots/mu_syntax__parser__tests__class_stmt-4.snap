---
source: crates/syntax/src/parser/tests.rs
expression: errors
---
[31merror[0m: expected `pass`
[34m>[0m code:1
[34m|[0m 
[34m|[0m class T: [31;4ma[0m = b
[34m|[0m 


